daml-foundations/daml-ghc/src/DA/Daml/GHC/Compiler/Convert.hs

```diff
@@ -990,66 +999,128 @@ mkProjBindings env recExpr recTyp vsFlds e =
 -- Coercions induced by newtypes can also occur deeply nested in function
 -- types or forall quantifications. We handle those cases by recursion into
 -- all sub-coercions.
-convertCast :: Env -> LF.Expr -> Coercion -> ConvertM LF.Expr
-convertCast env expr0 co0 = evalStateT (go expr0 co0) 0
+convertCoercion :: Env -> Coercion -> StateT Int ConvertM (LF.Expr, LF.Expr)
+convertCoercion env co
+    | isReflCo co = do
+        s' <- lift $ convertType env s
+        t' <- lift $ convertType env t
+        x <- mkLamBinder
+        let lamTo = ETmLam (x, s') (EVar x)
+            lamFrom = ETmLam (x, t') (EVar x)
+        pure (lamTo, lamFrom)
+    | Just (xCo, yCo) <- splitFunCo_maybe co = do
+        let Pair a a' = coercionKind xCo
+        (aTo, aFrom) <- convertCoercion env xCo
+        (bTo, bFrom) <- convertCoercion env yCo
+        sLf <- lift $ convertType env s
+        tLf <- lift $ convertType env t
+        aLf <- lift $ convertType env a
+        a'Lf <- lift $ convertType env a'
+        f <- mkLamBinder
+        g <- mkLamBinder
+        x <- mkLamBinder
+        y <- mkLamBinder
+        let lamTo =
+                ETmLam
+                    (f, sLf)
+                    (ETmLam
+                         (x, a'Lf)
+                         (bTo `ETmApp` (EVar f `ETmApp` (aFrom `ETmApp` EVar x))))
+        let lamFrom =
+                ETmLam
+                    (g, tLf)
+                    (ETmLam
+                         (y, aLf)
+                         (bFrom `ETmApp` (EVar g `ETmApp` (aTo `ETmApp` EVar y))))
+        pure (lamTo, lamFrom)
+    | Just (aGhc, k_co, co') <- splitForAllCo_maybe co
+    , isReflCo k_co -- TODO (drsk) is this restriction needed?
```

Comment:
This is a sanity check. We don't want to get involved in the business of kind coercions. All ways I can imagine how they pop up just don't work in DAML.